File: tabulation/kotlin/countConstruct.kt
fun countConstruct(target: String, wordBank: Array<String>): Int {
    val table: Array<Int> = Array(target.length + 1) { 0 }
    table[0]++

    for (i: Int in (0..target.length))
        for (word: String in wordBank)
            if (i + word.length <= target.length && target.substring(i, i + word.length) == word)
                table[i + word.length] += table[i]

    return table[target.length]
}

fun main(): Unit {
    println(countConstruct("purple", arrayOf("purp", "p", "ur", "le", "purpl")))
    println(countConstruct("javascript", arrayOf("java", "scr", "it", "javascr", "script")))
    println(countConstruct("aaaaaaaaaaaaaaaaaaa", arrayOf("a", "aa", "aaa", "aaaa", "aaaaa", "aaaaaa", "aaaaaaa", "aaaaaaaa", "aaaaaaaaa", "aaaaaaaaaa")))
    println(countConstruct("abcdef", arrayOf("a", "ab", "bcd", "f", "def", "bc")))
    println(countConstruct("skateboard", arrayOf("ska", "te", "boar", "ate", "sk", "at", "bo")))
    println(countConstruct("abcdef", arrayOf("a", "ab", "bcd", "def", "b", "c")))
}

File: tabulation/javascript/countConstruct.js
function countConstruct(target, wordBank) {
    const table = Array(target.length + 1).fill(0);
    table[0]++

    for (let i = 0; i <= target.length; i++)
        for (let word of wordBank)
            if (target.substr(i, word.length) === word)
                table[i + word.length] += table[i]

    return table[target.length]
}

console.log(countConstruct("purple", ["purp", "p", "ur", "le", "purpl"]))
console.log(countConstruct("javascript", ["java", "scr", "it", "javascr", "script"]))
console.log(countConstruct("aaaaaaaaaaaaaaaaaaa", ["a", "aa", "aaa", "aaaa", "aaaaa", "aaaaaa", "aaaaaaa", "aaaaaaaa", "aaaaaaaaa", "aaaaaaaaaa"]))
console.log(countConstruct("abcdef", ["a", "ab", "bcd", "f", "def", "bc"]))
console.log(countConstruct("skateboard", ["ska", "te", "boar", "ate", "sk", "at", "bo"]))
console.log(countConstruct("abcdef", ["a", "ab", "bcd", "def", "b", "c"]))